### PR TITLE
Retry oEmbed cache after redirect

### LINKED ISSUE #31308: Ubuntu 22.04.4 LTS installation mastodon v4.2.10 video preview on YouTube
### Steps to reproduce the problem

Ubuntu 22.04.4 LTS installation mastodon v4.2.10 video preview on YouTube

### Expected behaviour

11

### Actual behaviour

11

### Detailed description

111

### Mastodon instance

1111

### Mastodon version

111

### Browser name and version

11

### Operating system

11

### Technical details

111

### PR DESCRIPTION
**Pull request by c960657**
When a status contains a URL that redirects to a different URL, we currently only use the oEmbed cache to fetch the oEmbed data for the original URL. If the original URL has no cache entry, or the oEmbed endpoint returns no data for the original URL, and the HTML of the destination URL for some reason contains no oEmbed HTML tags (e.g. due to a CAPTCHA), it might be worth checking the oEmbed cache for destination URL. 

Why would the HTML sometimes not contain any oEmbed tags? This seems weird, but according to the findings by @alexandru in https://github.com/mastodon/mastodon/issues/31308#issuecomment-2333920424 it happens when YouTube's bot detection is triggered.

I think this might fix #31308 (and duplicates #31376 and #31462), though I cannot reproduce the issue myself.

### CODE REVIEW

**Comment by c960657** [comment]
I cannot reproduce the issue, but based on https://github.com/mastodon/mastodon/issues/31308#issuecomment-2333920424, I think this is what happens:

The user posts a status with a link to https://youtu.be/dQw4w9WgXcQ. That URL redirects to https://www.youtube.com/watch?v=dQw4w9WgXcQ&feature=youtu.be. When Mastodon fetches that link, it triggers YouTube's bot detection, and in that case the HTML page does not contain contain any HTML tags, so the `service.call(@url, html: html)` returns nil. But if the oEmbed endpoint URL for YouTube, https://www.youtube.com/oembed?format=json&url={url}, is cached, we can pass https://www.youtube.com/watch?v=dQw4w9WgXcQ&feature=youtu.be to that endpoint and get the oEmbed data.

**Comment by c960657** [comment]
It was mostly the output.html linked from that comment, I was referring to. If you open it in a browser, you can see how the page looks when bot detection is triggered. Note that the page contains no `<link rel="alternate" type="application/json+oembed" ...>` tag.

The issue is less likely to occur, if the status links to www.youtube.com, because that allows using the oEmbed cache and thus bypasses the bot detection, assuming that bot detection only affects the HTML page and not the oEmbed endpoint.

Currently, if the status link uses the YouTube short-linker, https://youtu.be, the oEmbed cache is not used, so the short-links are more likely to be affected by this issue.

**Comment by oneiros** [comment]
Do not get me wrong, the changes you propose seem to be an improvement, just not for this particular issue.

> because that allows using the oEmbed cache

The cache expires in 24h and this problem has been going on for weeks now, so I would not expect any relief from these changes :frowning_face: 

> ... and I fail to see why a redirect would be involved.

Well, there _could_ have been a redirect, but I just checked with `curl` on a hetzner machine and there is none.

So, again, aside from my remarks in the code, I believe these changes might be benificial, but I do not see how they would improve the youtube situation.

**Review by ClearlyClaire — COMMENTED** [review]
I can't see either how it could help with the YouTube situation.

The method is a bit hard to follow because of `html` possibly mutating `@url`, but that is not that PR's fault, it's a shortcoming of the existing code.

I need to think a bit more about using a cached OEmbed endpoint thing for the redirected-to URL, but at first glance it makes sense.

**Comment by c960657** [comment]
Yes, you are right.

This PR only improves the situation for links that redirect, such as short-links via youtu.be.

According to e.g. https://github.com/mastodon/mastodon/issues/31308#issuecomment-2322609303, links via youtu.be are currently less likely to work than direct links to www.youtube.com, possible because the latter can use the oEmbed cache. With this PR, both are equally likely to work, depending on whether the oEmbed cache for www.youtube.com is populated.

It is my impression that it works as long as the oEmbed cache is populated. And based on the preview cards in my own feed it is my impression that YouTube's bot detection does not kick in every single time, so there is a chance that the oEmbed cache will be populated now and then. So perhaps the issue can be mitigated further by extending the cache lifetime or making some more advanced expire mechanism that allows using a stale value.

**Comment by mjankowski** [comment]
It's been about two years since last update here ...

Is this blocked on a code concern about the cached endpoint thing?

Has the youtube-related stuff progressed at all in other PRs?

**Comment by oneiros** [comment]
> Has the youtube-related stuff progressed at all in other PRs?

I believe that more importantly it has been fixed on youtube's end (at least I have not seen the broken preview in a *long* time and it has been about equally long since I have seen any complaints about it).
